File: motifapalooza.pl
#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Std;

my $SEQS = 10;
my $LEN = 100;
my $FREQ = 0.9;
my $MPS  = 1;
my $PA   = 0.25;
my $PC   = 0.25;
my $PG   = 0.25;
my $PT   = 0.25;

my $usage = "
usage: $0 [options] <motif file>
options:
  -n <int>   number of sequences to generate [$SEQS]
  -l <int>   length of sequences to generate [$LEN]
  -f <float> frequency of sequences with motifs [$FREQ]
  -m <int>   number of motifs per sequence [$MPS]
  -b         include motifs on both strands [off]
  -a <float> background probablility of A [$PA]
  -c <float> background probablility of C [$PC]
  -g <float> background probablility of G [$PG]
  -t <float> background probablility of T [$PT]
";

our ($opt_n, $opt_l, $opt_f, $opt_m, $opt_b,
	$opt_a, $opt_c, $opt_g, $opt_t);
getopts('n:l:f:m:ba:c:g:t:');
$SEQS = $opt_n if $opt_n;
$LEN  = $opt_l if $opt_l;
$FREQ = $opt_f if $opt_f;
$MPS  = $opt_m if $opt_m;
$PA   = $opt_a if $opt_a;
$PC   = $opt_c if $opt_c;
$PG   = $opt_g if $opt_g;
$PT   = $opt_t if $opt_t;
my $BOTH_STRANDS = $opt_b;

die $usage unless @ARGV == 1;

my $motif = read_jaspar($ARGV[0]);
my $mlen = @$motif;
for (my $i = 0; $i < $SEQS; $i++) {
	my $seq = generate_seq($LEN, $PA, $PC, $PG, $PT);
	my @info;
	if (rand(1) < $FREQ) {
		for (my $m = 0; $m < $MPS; $m++) {
			my $site = generate_site($motif);
			my $strand = '+';
			if ($BOTH_STRANDS and rand(1) < 0.5) {
				$site =~ tr/ACGTacgt/TGCAtgca/;
				$site = reverse $site;
				$strand = '-';
			}
			my $pos = int(rand($LEN - $mlen +1));
			push @info, "$pos$strand";
			substr($seq, $pos, $mlen) = $site;
		}
	}
	print ">seq-$i motifs: @info\n", "$seq\n";
}

sub generate_seq {
	my ($len, $pa, $pc, $pg, $pt) = @_;
	my $seq = "";
	for (my $i = 0; $i < $len; $i++) {
		my $r = rand(1);
		if    ($r < $pa)         {$seq .= 'a'}
		elsif ($r < $pa+$pc)     {$seq .= 'c'}
		elsif ($r < $pa+$pc+$pg) {$seq .= 'g'}
		else                     {$seq .= 't'}
	}
	return $seq;
}

sub generate_site {
	my ($motif) = @_;
	
	my $site = "";
	for (my $i = 0; $i < @$motif; $i++) {
		my $tA = $motif->[$i]{A};
		my $tC = $tA + $motif->[$i]{C};
		my $tG = $tC + $motif->[$i]{G};
		my $r = rand(1);
		if    ($r < $tA) {$site .= 'A'}
		elsif ($r < $tC) {$site .= 'C'}
		elsif ($r < $tG) {$site .= 'G'}
		else             {$site .= 'T'}
	}
	
	return $site;
}

sub read_jaspar {
	my ($file) = @_;
	
	open(my $fh, $file) or die;
	my $head = <$fh>;
	my ($id, $name) = $head =~ /^>(\S+)\s+(\S+)/;
	my @motif;
	for (my $i = 0; $i < 4; $i++) {
		my $line = <$fh>;
		$line =~ s/[\[\]]//g;
		my ($nt, @val) = split(/\s+/, $line);
		for (my $j = 0; $j < @val; $j++) {
			$motif[$j]{$nt} = $val[$j];
		}
	}
	
	for (my $i = 0; $i < @motif; $i++) {
		my $total = 0;
		foreach my $nt (keys %{$motif[$i]}) {$total += $motif[$i]{$nt}}
		foreach my $nt (keys %{$motif[$i]}) {$motif[$i]{$nt} /= $total}
	}
	
	return \@motif;
}

File: motifapalooza.py
#!/usr/bin/env python3
import argparse
import random

# setup
parser = argparse.ArgumentParser(
	description='Generates random sequence and embeds known fungal TF binding motifs.')
# required arguments
parser.add_argument('--jasparfile', required=True, type=str,
	metavar='<str>', help='jaspar file')
parser.add_argument('--numseq', required=False, type=int, default=10,
	metavar='<int>', help='number of sequences to generate [%(default)i]')
parser.add_argument('--seqlen', required=False, type=int, default=100,
	metavar='<int>', help='length of sequences to generate [%(default)i]')
parser.add_argument('--mps', required=False, type=int, default=1,
	metavar='<int>', help='number of motifs per seqeunce [%(default)i]')
parser.add_argument('--freq', required=False, type=float, default=0.9,
	metavar='<float>', help='optional floating point argument [%(default)f]')	
parser.add_argument('--PA', required=False, type=float, default=0.25,
	metavar='<float>', help='optional floating point argument [%(default)f]')
parser.add_argument('--PC', required=False, type=float, default=0.25,
	metavar='<float>', help='optional floating point argument [%(default)f]')
parser.add_argument('--PG', required=False, type=float, default=0.25,
	metavar='<float>', help='optional floating point argument [%(default)f]')
parser.add_argument('--PT', required=False, type=float, default=0.25,
	metavar='<float>', help='optional floating point argument [%(default)f]')
# switches
parser.add_argument('--bothstrands', action='store_true',
	help='on/off switch')
# finalization
arg = parser.parse_args()

def read_JASPAR(jasparfile):
	motif = []
	linenum = 0
	nt = ['A', 'C', 'G', 'T']
	with open(arg.jasparfile) as jf:
		for line in jf.readlines():
			if line.startswith('>'): continue
			line = line.split()	
			for i in range(2,len(line)-1):
				if len(motif) < (len(line)-3): motif.append({})
				motif[i-2][nt[linenum]] = int(line[i])
			linenum+= 1 
	for i in range(len(motif)):
		total = 0 
		for nt in motif[i]: total += motif[i][nt]
		for nt in motif[i]: motif[i][nt] /= total
	return motif 
		
		
def generate_site(motif):
	motifseq = []
	total = 0
	for i in range(0,len(motif)):
		pA = motif[i]['A']
		pC = motif[i]['C']
		pG = motif[i]['G']
		pT = motif[i]['T']
		r = random.random()
		if r < pA:               motifseq += 'A'
		elif r < (pA + pC):      motifseq += 'C'
		elif r < (pA + pC + pG): motifseq += 'G' 
		else:                    motifseq += 'T'
	return motifseq
				
	
def generate_seq(numseq, seqlen, PA, PC, PG, PT):
	seq = []
	for i in range(0,arg.seqlen):
		r  = random.random()
		if r < PA:             seq += 'a'
		elif r < PA + PC:      seq += 'c'
		elif r < PA + PC + PG: seq += 'g'
		else:                  seq += 't'
	return seq
		
		
motif = read_JASPAR(arg.jasparfile)
for i in range(0, arg.numseq):
	seq = (generate_seq(arg.numseq, arg.seqlen, arg.PA, arg.PC, arg.PG, arg.PT))
	r = random.random()
	places = []
	if r < arg.freq:
		for j in range(0,arg.mps):
			site = generate_site(motif)
			assert(len(motif)== len(site))
			place = random.randint(0, len(seq)-len(motif))
			places.append(f'{place}+')
			for k in range(0,len(site)):
				seq[place+k] = site[k]
	print(f'>seq-{i} {places}')
	print(''.join(seq))

	
		
		
		
	


